apps/src/lib/node/ledger/mod.rs

```diff
@@ -609,27 +614,63 @@ fn start_tendermint(
 ///
 /// An oracle is also returned, along with its associated channel,
 /// for receiving Ethereum events from `geth`.
-async fn start_ethereum_node(
+async fn maybe_start_ethereum_oracle(
+    config: &config::Ledger,
+    abort_sender: oneshot::Sender<()>,
+) -> (Option<mpsc::Receiver<EthereumEvent>>, task::JoinHandle<()>) {
+    if !matches!(config.tendermint.tendermint_mode, TendermintMode::Validator) {
+        return (None, spawn_dummy_task(()));
+    }
+
+    let ethereum_url = config.ethereum.oracle_rpc_endpoint.clone();
+
+    // Start the oracle for listening to Ethereum events
+    let (eth_sender, eth_receiver) = mpsc::channel(ORACLE_CHANNEL_BUFFER_SIZE);
+    let oracle = match config.ethereum.mode {
+        ethereum::Mode::Managed | ethereum::Mode::Remote => {
+            ethereum_node::oracle::run_oracle(
+                ethereum_url,
+                eth_sender,
+                abort_sender,
+            )
+        }
+        ethereum::Mode::EventsEndpoint => {
+            ethereum_node::test_tools::events_endpoint::serve(
+                eth_sender,
+                abort_sender,
+            )
+        }
+        ethereum::Mode::Off => spawn_dummy_task(()),
+    };
+
+    (Some(eth_receiver), oracle)
+}
+
+/// Launches a new task managing a `geth` process into the asynchronous
+/// runtime, and returns its [`task::JoinHandle`].
+///
+/// An oracle is also returned, along with its associated channel,
+/// for receiving Ethereum events from `geth`.
+async fn maybe_start_geth(
```

Comment:
tbh, I'm not a big fan of returning those dummy tasks. a new refactor could be in order to return `Option<JoinHandle<T>>` values from some of these futures. then, we could use `futures::try_join_all`, on a `Vec` we collected with the flattening of all futures we want to wait on, from `Option<F>` for some `F: Future`. we'd probably need `F` to be `Pin<Box<dyn Future<Output = ShellResult<()>>>>` I guess

Reply:
Yeah we need to refactor this further - dummy tasks should eventually no longer required. Linking to https://github.com/anoma/namada/issues/459. I briefly tried something like the `Vec` approach you mention but we need to update all our tasks to return the same type (e.g. `JoinHandle<Result<()>>`) which they don't currently do, also we run the shell in a regular thread rather than a tokio task for some reason.

Reply:
the shell could potentially execute on a blocking tokio task I think, which can also have a `JoinHandle` of its own.
https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html